processing/src/main/java/org/apache/druid/query/CacheStrategy.java

```diff
@@ -48,21 +50,27 @@
   boolean isCacheable(QueryType query, boolean willMergeRunners);
 
   /**
-   * Computes the cache key for the given query
+   * Computes the per-segment cache key for the given query. Because this is a per-segment cache key, it should only
+   * include parts of the query that affect the results for a specific segment (i.e., the results returned from
+   * {@link QueryRunnerFactory#createRunner}).
    *
-   * @param query the query to compute a cache key for
+   * @param query the query to be cached
    *
-   * @return the cache key
+   * @return the per-segment cache key
    */
   byte[] computeCacheKey(QueryType query);
 
   /**
-   * Computes the result level cache key for the given query.
-   * Some implementations may include query parameters that might not be used in {@code computeCacheKey} for same query
+   * Computes the result-level cache key for the given query. The result-level cache will tack on datasource and
```

Comment:
I was going to file a bug about result-level cache key computation does not taken interval details into account until this comes along. If this is intentional, it still not clear about how the result-level cache "tacking on datasource and interval details". Maybe it's better to add more docs on this part?

Reply:
It does it in sort of an easy-to-miss way: the key actually doesn't contain datasource and interval details, but the _value_ does, and the ResultLevelCachingQueryRunner will only use the cached value if they match. Check out ResultLevelCachingQueryRunner to see how it works.

Fwiw, I think it would be good to modify ResultLevelCachingQueryRunner to make this part of the key, not part of the value, so it can cache results for queries that differ in interval but are the same otherwise. I don't have plans to do this personally though, at least not right now.